### PR TITLE
Improve parity between nvfuser and TE's quantization to NVFP4

### PR DESCRIPTION
**Pull request by protonu**
This makes minor modifications to the runtime function for NVFP4 quantization to make the quantized match the quantized values from Transformer Engine. This updates tests to reflect the new runtime function.

### CODE REVIEW

**Comment by protonu** [comment]
!test

**Comment by protonu** [comment]
!test

File: tests/python/direct/test_narrow_precision.py [hunk 1]
```diff
@@ -443,17 +450,7 @@ def fusion_baseline(fd: FusionDefinition) -> None:
...
-    max_diff = torch.max(abs_diff)
-    assert max_diff <= 10.0, f"Max difference {max_diff:.4f} exceeds threshold of 10.0"
-
-    # Check that large differences (> 5.0) are rare (< 10% of elements)
-    large_diff_count = torch.count_nonzero(torch.gt(abs_diff, 5.0))
-    large_diff_ratio = large_diff_count / abs_diff.numel()
-    assert (
-        large_diff_ratio < 0.1
-    ), f"Large diff ratio {large_diff_ratio:.2%} exceeds 10% threshold"
+    torch.testing.assert_close(outputs[0], outputs_baseline[0], atol=1e-2, rtol=1e-2)
```
Comment (jjsjann123) [inline tests/python/direct/test_narrow_precision.py:453]: I didn't see any changes to the actual math for global scaling factor and how scale is applied on the data tensor.

So this indeed is just numerics?! 😰  sorry for pushing you through this....

File: tests/python/direct/test_narrow_precision.py [hunk 2]
```diff
@@ -146,12 +144,23 @@ def compute_nvfp4_global_scale(tensor):
...
 
     # Handle edge cases: zero input or invalid scale
-    invalid_mask = (amax == 0.0) | (x_global_scale == 0.0)
-    x_global_scale = torch.where(invalid_mask, 1.0, x_global_scale)
+    if x_global_scale == torch.tensor(
+        0.0, device=x_global_scale.device, dtype=torch.float32
+    ):
+        x_global_scale = torch.tensor(
+            1.0, device=x_global_scale.device, dtype=torch.float32
+        )
```
Comment (jjsjann123) [inline tests/python/direct/test_narrow_precision.py]: nitpick on computing global scale, do we need tensor operations or it's fine to move the amax value to cpu and use vanilla python math on it?
Also, the comment about `if` on tensor from greptile looks very legit!

**Review by jjsjann123 — COMMENTED** [review]
Yay validation improvement!

File: tests/python/direct/test_narrow_precision.py (hunk 2 above)
Comment (protonu) [reply tests/python/direct/test_narrow_precision.py]: We do need a o-d tensor as input to the fusion. That's why I kept these as tensor operations.

**Comment by protonu** [comment]
!test

File: tests/python/direct/test_narrow_precision.py [hunk 3]
```diff
@@ -146,12 +144,21 @@ def compute_nvfp4_global_scale(tensor):
...
-    x_global_scale = torch.clamp(x_global_scale, max=torch.finfo(torch.float32).max)
+    x_global_scale = torch.div(FLOAT8_E4M3_MAX * FLOAT4_E2M1_MAX, amax)
+    x_global_scale = torch.min(
+        x_global_scale,
+        torch.tensor(
+            torch.finfo(torch.float32).max,
+            device=x_global_scale.device,
+            dtype=torch.float32,
+        ),
+    )
```
[the remark this answers is not in the document]
Comment (protonu) [reply tests/python/direct/test_narrow_precision.py]: No, going with min to match TE's code.

File: tests/python/direct/test_narrow_precision.py [hunk 4]
```diff
@@ -146,12 +144,21 @@ def compute_nvfp4_global_scale(tensor):
         Global scale as (FLOAT8_E4M3_MAX * FLOAT4_E2M1_MAX) / max(abs(tensor))
     """
     amax = torch.max(torch.abs(tensor)).to(torch.float32)
-    x_global_scale = (FLOAT8_E4M3_MAX * FLOAT4_E2M1_MAX) / amax
-    x_global_scale = torch.clamp(x_global_scale, max=torch.finfo(torch.float32).max)
+    x_global_scale = torch.div(FLOAT8_E4M3_MAX * FLOAT4_E2M1_MAX, amax)
```
[the remark this answers is not in the document]
Comment (protonu) [reply tests/python/direct/test_narrow_precision.py]: Going with torch.div to match TE's code - though this probably doesn't make a difference.

File: runtime/block_quantization_kernels.cu [hunk 5]
```diff
@@ -115,23 +115,24 @@ __device__ void block_quantize_to_nvfp4(
...
 
   __e4m3 clamped_max_fp8 = __float2e4m3(scaled_max);
 
   // Convert back from FP8 to float using __e4m32float
   scaled_max = __e4m32float(clamped_max_fp8);
 
   if constexpr (USE_GLOBAL_SCALE) {
-    scaled_max = global_scale[0] / scaled_max;
+    auto decode_scale = 1.0f / global_scale[0];
+    scaled_max = fminf(1.0f / (scaled_max * decode_scale), float_max);
```
[the remark this answers is not in the document]
Comment (protonu) [reply runtime/block_quantization_kernels.cu:133]: Avoiding a division in this code - to match TE.

**Comment by protonu** [comment]
!test

File: tests/python/direct/test_narrow_precision.py (hunk 2 above)
Comment (jjsjann123) [reply tests/python/direct/test_narrow_precision.py]: with 0-d tensor, we can do `tensor.item()` to get its value out. I was wondering if that would allow us to switch to python math instead.

**Review by jjsjann123 — APPROVED** [review]
stamping with the comment on scalar tensor handling in global scaling factor.

**Comment by protonu** [comment]
I don't think it's possible to get bitwise parity over a large spread of inputs sizes - I think TE may dispatch to different kernel with slightly different numerics.

**Comment by protonu** [comment]
!test

**Comment by protonu** [comment]
!test

**Comment by protonu** [comment]
!test

**Comment by protonu** [comment]
!test
